### PR TITLE
[RESOLVED] system config error

### LINKED ISSUE #12566: [mesheryctl] `system config gke`: Error generating config with null input
_See [mesheryctl Command Tracker](https://bit.ly/3dqXy1q) for current status of commands._

### Current Behavior
See logs below.

### Screenshots/Logs
```
$ mesheryctl system config gke
Configuring Meshery to access GKE...
Creating a service account in default namespace: sa-meshery-ikzymhhd
serviceaccount/sa-meshery-ikzymhhd created
[clusterrolebinding.rbac.authorization.k8s.io/sa-meshery-ikzymhhd](http://clusterrolebinding.rbac.authorization.k8s.io/sa-meshery-ikzymhhd) created
Getting secret of service account sa-meshery-ikzymhhd on default
jq: error (at <stdin>:11): Cannot iterate over null (null)
Error generating config:exit status 5
```

### Environment

- Host OS: Windows
- Meshery Client Version: stable-v0.7.133

---

### Contributor [Guides](https://docs.meshery.io/project/contributing) and [Handbook](https://layer5.io/community/handbook)
- 📚 [mesheryctl Contributing Guide](https://github.com/meshery/meshery/blob/master/mesheryctl/README.md)
- 📺 [Beginner's guide to contributing to Meshery and mesheryctl](https://youtu.be/hh_kFLZx3G4)
- ⌨️ [mesheryctl Command Tracker](https://docs.google.com/spreadsheets/d/1q63sIGAuCnIeDs8PeM-0BAkNj8BBgPUXhLbe1Y-318o/edit#gid=0)
- ⌨️ [Meshery CLI Commands and Documentation](https://docs.google.com/document/d/1xRlFpElRmybJ3WacgPKXgCSiQ2poJl3iCCV1dAalf0k/edit#heading=h.5fucij4hc5wt)
- 🙋🏾🙋🏼 Questions: [Discussion Forum](http://discuss.meshery.io) and [Community Slack](https://slack.meshery.io)

### PR DESCRIPTION
**Pull request by dev-tnsq**
### **[mesheryctl] Fix for system config GKE null input error**

This PR fixes #12566.

**Description**
This PR addresses an issue in the `GenerateConfigGKE` function in `scripts.go` where a null input for the GKE configuration led to a failure in retrieving service account secrets. 

**Key Changes**
- Added a 5-second delay after creating the service account to ensure proper initialization.
- Implemented validation checks for:
  - Service account existence
  - Non-empty secret name
- Enhanced error handling and improved error messages for better troubleshooting.
- Improved the readability so that It can be easily understandable

- [x]  Yes, I signed my commits.

## **Screenshots: Before and After the Fix**

**Before:**
Screenshot 2024-11-10 at 4 22 53 AM
**After:**
Screenshot 2024-11-14 at 4 03 25 AM

### CODE REVIEW

**Comment by Tanishq1604** [comment]
@lekaf974, could you please help me with this? I'm not sure why the integration tests are failing, especially since they run without any issues on my local setup. Any insights into what might be causing this difference?

Thanks in advance!

File: mesheryctl/pkg/utils/scripts.go [hunk 1]
```diff
@@ -6,118 +6,250 @@ import (
...
-			echo -e "\\nMissing required utility: 'awk'. Please install 'awk' and try again."
-			exit;
-		fi
-		if ! command_exists tail ; then
-			echo -e "\\nMissing required utility: 'tail'. Please install 'tail' and try again."
-			exit;
-		fi
+
+	if err := cfg.validate(); err != nil {
+		return fmt.Errorf("validation failed: %w", err)
```
Comment (lekaf974) [inline mesheryctl/pkg/utils/scripts.go:26]: wondering if it should return mesheryctl err

Comment (Tanishq1604) [reply mesheryctl/pkg/utils/scripts.go:26]: I used `fmt.Errorf` to improve output formatting. Let me know if your opinion , I'm happy to adjust as needed.

Comment (Tanishq1604) [reply mesheryctl/pkg/utils/scripts.go:26]: Screenshot 2024-11-15 at 12 50 11 PM
i guess using this should be better ??

**Comment by vishalvivekm** [comment]
@Tanishq1604 
Let's discuss this during the Meshery Development call today at 9:00 AM CT | 8:30 PM IST. Adding it as an agenda item to the [meeting minutes](https://docs.google.com/document/d/1Dg4Uz9rGTuGHmYZwIt47z1j5x3LmnpxD1M1SgypAkQ0/edit?pli=1#) || [meeting link](https://meet.layer5.io/meshery).

**Comment by lekaf974** [comment]
thx @Tanishq1604 LGTM, you could take avantage  to verify if there is a test on `mesheryctl system config gke` command and if not we could add it ?

File: mesheryctl/pkg/utils/scripts.go [hunk 2]
```diff
@@ -6,118 +6,250 @@ import (
...
+// GKEConfig holds configuration for GKE script generation
+
+type GKEConfig struct {
+	ConfigPath string
+	SAName     string
+	Namespace  string
+}
+
+// GenerateConfigGKE generates kubernetes config for GKE
+func GenerateConfigGKE(configPath, SAName, namespace string) error {
```
Comment (lekaf974) [inline mesheryctl/pkg/utils/scripts.go:18]:
```suggestion
func GenerateConfigGKE(configPath, serviceAccountName, namespace string) error {
```

File: mesheryctl/pkg/utils/scripts.go [hunk 3]
```diff
@@ -6,118 +6,250 @@ import (
...
-		echo -e "\\nCreating a service account in ${NAMESPACE} namespace: ${SERVICE_ACCOUNT_NAME}"
-		kubectl create sa "${SERVICE_ACCOUNT_NAME}" --namespace "${NAMESPACE}"
-		kubectl create clusterrolebinding "${SERVICE_ACCOUNT_NAME}" --clusterrole=cluster-admin --serviceaccount=default:"${SERVICE_ACCOUNT_NAME}" --namespace "${NAMESPACE}"
+
+	return cfg.executeScript()
+}
+
+func (c *GKEConfig) validate() error {
+	if c.ConfigPath == "" || c.SAName == "" || c.Namespace == "" {
+		return fmt.Errorf("configPath, SAName, and namespc are required")
```
Comment (lekaf974) [inline mesheryctl/pkg/utils/scripts.go:34]: Should create a mesherytl error and return it

you can check https://github.com/lekaf974/meshery/blob/master/mesheryctl/internal/cli/root/design/error.go as example and read the error guide https://docs.meshery.io/project/contributing/contributing-error


ps: Can you proceed kike this for all errors returned

File: mesheryctl/pkg/utils/scripts.go (hunk 2 above)
Comment (Tanishq1604) [reply mesheryctl/pkg/utils/scripts.go:18]: @lekaf974 ,i guess in the internal where is this function is being used they are passing it as SAName 
Screenshot 2024-11-21 at 8 16 59 PM

**Comment by Tanishq1604** [comment]
@hamza-mohd , if u are trying to change the SA Name then it is being used in the whole file and also as I said that before ,In the mesheryctl in internal in config.go it is being passed as SAName

File: mesheryctl/pkg/utils/scripts.go (hunk 2 above)
Comment (lekaf974) [reply mesheryctl/pkg/utils/scripts.go]: Ok get it back `SAName` if it is already used like that in the code

File: mesheryctl/pkg/utils/scripts.go [hunk 4]
```diff
@@ -6,118 +6,252 @@ import (
...
+	fmt.Println("\nKubernetes API")
+	fmt.Println("--------------")
+
+	// Check if client can be initialized
+	cmd := exec.Command("kubectl", "cluster-info")
+	cmd.Stdout = nil
+	cmd.Stderr = nil
+
+	if err := cmd.Run(); err != nil {
+		fmt.Println("!! cannot connect to Kubernetes API server")
```
Comment (lekaf974) [inline mesheryctl/pkg/utils/scripts.go:89]:
```suggestion
```
There error message should be enough by itself

File: mesheryctl/pkg/utils/scripts.go [hunk 5]
```diff
@@ -6,118 +6,252 @@ import (
...
+	// Check API query
+	cmd = exec.Command("kubectl", "get", "nodes")
+	if err := cmd.Run(); err != nil {
+		fmt.Println("!! cannot query Kubernetes API")
+		return ErrKubernetesQuery(fmt.Errorf("failed to query Kubernetes API: %w", err))
 	}
+	fmt.Println("✓ can query Kubernetes API")
+
+	fmt.Println("\n--------------")
+	fmt.Println("--------------")
```
Comment (lekaf974) [inline mesheryctl/pkg/utils/scripts.go:103]:
```suggestion
```

Comment (lekaf974) [reply mesheryctl/pkg/utils/scripts.go:103]: Let's keep output cleat to have a better User experience

File: mesheryctl/pkg/utils/scripts.go [hunk 6]
```diff
@@ -6,118 +6,252 @@ import (
...
+	}
+	return nil
+}
+
+func (c *GKEConfig) checkPrerequisites() error {
+	var missingCommands []string
+	requiredCommands := []string{"jq", "base64", "awk", "tail"}
+
+	fmt.Println("\nRequired Commands")
+	fmt.Println("--------------")
```
Comment (lekaf974) [inline mesheryctl/pkg/utils/scripts.go:46]:
```suggestion
	fmt.Println("\nRequired Commands")
	fmt.Println("--------------")
```

**Comment by lekaf974** [comment]
branch needs to be updated as well

File: mesheryctl/pkg/utils/scripts.go (hunk 4 above)
Comment (Tanishq1604) [reply mesheryctl/pkg/utils/scripts.go:89]: signed off

**Review by lekaf974 — COMMENTED** [review]
LGTM, @alphaX86 can you have a second look ?

**Comment by alphaX86** [comment]
> LGTM, @alphaX86 can you have a second look ?

ACK! 👍

File: mesheryctl/pkg/utils/scripts_test.go
Comment (alphaX86) [inline mesheryctl/pkg/utils/scripts_test.go:1]: @Tanishq1604 I see that you've created a test for this. Does this return proper test results?

Comment (Tanishq1604) [reply mesheryctl/pkg/utils/scripts_test.go:1]: @alphaX86 yes

**Review by lekaf974 — APPROVED** [review]
LGTM

**Review by alphaX86 — APPROVED** [review]

**Comment by alphaX86** [comment]
Thanks @Tanishq1604 for your work! This is good to merge

**Pull request merged**